### DISCUSSION TITLE
Disable React.Suspense for Async Selectors

### DISCUSSION THREAD

**Discussion by mrousavy** [facebookexperimental/Recoil#1472] [Q&A]
Hi all!

I'm using an async selector that fetches a fair amount of data, it takes about 2-4 seconds on my phone. This is called on the app start, and it feels like the entire app freezes because nothing gets rendered during that time.

I tried debugging this to find out what part of Recoil might have performance issues and blocks the entire JS thread, and one thing was particularily interesting to me was: My async selector always returns a value, it's never `undefined`.

So I browsed the Recoil documentation only to find out [that Recoil is designed with React.Suspense in mind](https://recoiljs.org/docs/guides/asynchronous-data-queries/#asynchronous-example), which might be the reason that my entire app looks like it freezes for a few seconds on app launch (and renders nothing).

I wonder if there is a way to not use React.Suspsense, and instead have a `defaultValue` (of `undefined` or `[]`) for async selectors?

> React.Suspense doesn't work with my current app architecture (React Native)

**Comment by drarmstr** [comment]
Yes, if you don't want to use Suspense you can use [`useRecoilStateLoadable()`](https://recoiljs.org/docs/api-reference/core/useRecoilStateLoadable), [`useRecoilValueLoadable()`](https://recoiljs.org/docs/api-reference/core/useRecoilValueLoadable), and/or [`noWait()`](https://recoiljs.org/docs/api-reference/utils/noWait). And manually handle the `loading` state.

**Reply by mrousavy** [reply]
Yea that's what I'm using right now, but slapping `noWait(...)` on to every async `get()` can get a bit annoying. Do you think a third parameter in the `get` function makes sense, like `getOrDefault(...)`?

**Reply by mrousavy** [reply]
Also @drarmstr quick question: Do selectors support caching/persisting (via effects) or do I have to implement that manually?

**Reply by drarmstr** [reply]
Could be ambiguous if the "orDefault" is referring to a loading state or an undefined or nullable value.  But, either way, it's totally reasonable to make a wrapper to compose functionality like this if you find it helpful:
```jsx
function useRecoilValueOrDefault(node, fallback) {
  const loadable = useRecoilValueLoadable(node);
  switch (loadable.state) {
    case 'hasValue':
      return loadable.contents;
    case 'hasError':
      throw loadable.contents;
    case 'loading':
      return fallback;
  }
}
```

**Reply by drarmstr** [reply]
Selectors support caching, yes.  Selectors do not have effects comparable to atom effects, though, as that could lead to dangerous inconsistencies between atoms and derived selector state.  But, you can use React effects or atom effects for persistence.

**Reply by mrousavy** [reply]
I've just experienced a bit more with `noWait`/`waitForNone`, and that isn't really a feasible solution. The state will never be `loading`, it is always `hasValue` because I return early (if my async selector calls another async selector with `noWait`, it might return 'loading', but then the "outer" selector returns early (undefined), causing this one to always be `hasValue`).

So I can't really have a clean "loading" state management here, unless I make every single selector return `loading` (a string value) instead of it's actual type when one of the dependencies is still loading. And that becomes an additional value to check in my `useRecoilValueLoadable` abstraction, and is not really safe and can collide with the same value.

So I guess I really need a way to _disable_ the React.Suspense behaviour and instead have a `defaultValue: undefined` for the `selector({ ... })` which will be returned if it is still evaluating ('loading').

**Reply by mrousavy** [reply]
@drarmstr here's an example:

```ts
  const TokenBalances = selector({
    key: 'token-balances',
    get: async ({ get }) => {
      const walletAddress = get(WalletAddress)
      return await getTokenBalancesForWallet(walletAddress)
    },
  })


 const MyTokens = selector<TokenWithBalance[]>({
    key: 'my-tokens',
    get: ({ get }) => {
      const balancesLoadable = get(noWait(TokenBalances))
      if (balancesLoadable.state !== 'hasValue') return [] // <-- this causes the 'MyTokens' selector to always have `hasValue` (never `loading` state)
      const balances = balancesLoadable.contents

      const tokensLoadable = get(
        // v this causes the 'MyTokens' selector to return even when no token-details have been fully loaded ([]). I can't use waitForAll here since this then freezes my entire UI due to React.Suspense.
        waitForNone(balances.map((b) => TokenDetailsWithPrice(b.tokenSymbol))),
      )
      const tokens = tokensLoadable
        .map((t) => (t.state === 'hasValue' ? t.contents : undefined))
        .filter(notEmpty)

      return tokens.map((t) => {
        const tokenBalance = balances.find((b) => b.tokenSymbol === t.symbol)
        const balance = tokenBalance?.balance ?? 0

        return {
          ...t,
          balance: {
            tokenAmount: balance,
            price: {
              currency: t.price.currency,
              price: t.price.price * balance,
            },
          },
        }
      })
    },
  })

```

As you can see, if the `TokenBalances` selector is still in a loading state, I return `[]` in the `MyTokens` selector, causing the `MyTokens` selector to always have a value (`hasValue` instead of `loading`).

Now I could return `"loading"` instead of `[]` here and handle that in my `useRecoilValueLoadable` abstraction, but that's a lot of complicated code (since I have to put that in every early-return) and that actually doesn't work for incremental/parallel loading (`waitForNone`).

My proposed solution would be: Have 2 more props in `selector({ ... })`:

* `defaultValue` - a value to return while it is still in `"loading"` state. Probably `undefined` in most cases
* `disableReactSuspense` - to disable the suspense behaviour and instead return `defaultValue` while the `get` function is still executing. Makes `defaultValue` prop required.


wdyt?

**Reply by mrousavy** [reply]
You're gonna hate me, but this is how I currently hotpatched it in my project:

```diff
diff --git a/node_modules/recoil/native/recoil.js b/node_modules/recoil/native/recoil.js
 index 93c9175..57d78e8 100644
 --- a/node_modules/recoil/native/recoil.js
 +++ b/node_modules/recoil/native/recoil.js
 @@ -4486,6 +4486,8 @@ function handleLoadable(loadable, recoilValue, storeRef) {
    if (loadable.state === 'hasValue') {
      return loadable.contents;
    } else if (loadable.state === 'loading') {
 +    // VERY dirty hotfix to avoid React.Suspense.
 +    return undefined;
      const promise = new Promise(resolve => {
        storeRef.current.getState().suspendedComponentResolvers.add(resolve);
      }); // $FlowFixMe Flow(prop-missing) for integrating with tools that inspect thrown promises @fb-only
```

I'm interested in hearing your thoughts on this "feature" before working on a PR to add it. Ideally, this would allow returning `defaultValue` (persisted cache) instead of `undefined`.

**Reply by drarmstr** [reply]
Not sure why the extra options for `selector()` would be necessary when you demonstrated how the same behavior can be accomplished for a specific selector with `noWait()`.

For the complete disabling of suspense instead of your patch of returning `undefined` when loading, you could use a wrapper as I have above.  Your patch is problematic because it would cause existing hooks to potentially return `undefined` when that is not necessarily a potential value for the type of the atom/selector which would be against the TypeScript/Flow return type for the hooks.

**Reply by mrousavy** [reply]
> Not sure why the extra options for `selector()` would be necessary when you demonstrated how the same behavior can be accomplished for a specific selector with `noWait()`.

1. Using `noWait()` everywhere makes a loadable's `.status` be `hasValue`.
2. Since the loadable's status is `hasValue`, I need the value (`.contents`) to be some sort of loading placeholder, e.g. the string `"loading"`.
3. Then, in my custom hook abstraction, I need to check if the `.status` is `loading` **or** if the `.status` is `hasValue` and the `.contents` is `"loading"` to return a loading state. As you can see already, it's becoming insanely complex.
4. For arrays, this gets even more complex. `waitForAll()` will enter the Suspense again. `waitForNone()` makes me check each element for the condition in point 3.

I'm not confident with putting all of that _insanely complex_ logic in my state management code which should be a simple API call.

**Reply by mrousavy** [reply]
> For the complete disabling of suspense instead of your patch of returning undefined when loading, you could use a wrapper as I have above.

Sorry what do you mean exactly? `useRecoilValueLoadable()` also enters the suspense and blocks UI.

**Reply by drarmstr** [reply]
Well, the point of the loading state is to be able to propagate it around without having to manually handle it. So, if you want to disable Suspense globally then you don't need to add that logic per-selector and just use something like the wrapper hook alive in your component instead. UseRecoilValueLoadable() will not trigger suspense for that component by returning Loadable<T> instead of type T.  The wrapper above would be easier to use by not having to handle Loadable<T>, but would return T | void or T | FallbackType instead.

**Reply by mrousavy** [reply]
Yes, but it's still hard to maintain. Every single async `get` in a Selector has to be handled with a `noWait` (or `waitForNone`) and returning a fallback type.

**Reply by drarmstr** [reply]
Only if you specifically want to handle that case for that selector.  If your goal is just not use Suspense then you don't need to do anything per-selector and can just use the different hook at the component render.

**Reply by CollinMonahanLab49** [reply]
So if I have, as follows the docs,

```
<RecoilRoot>
  <React.Suspense fallback={<LoadingScreen />}>
    <AppOfManyFeatures />
  </React.Suspense>
</RecoilRoot>
```

Please verify - if any component is waiting a value from Recoil, then the whole `AppOfManyFeatures` is not rendered.

Can it be more granular than that?  Suspense and specifically-designed Loading visuals at the leaves of the component tree?

**Reply by drarmstr** [reply]
Sure, it can be more granular. Just put a <Suspense> deeper in the tree. A pending hook will bubble up to the closet ancestor <Suspense> component.
